File: combi_examples/combi_classic.py
# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at 
# sgpp.sparsegrids.org

'''
Python example for the combination technique. Be sure that the folder containing
pysgpp.py can be found in the PYTHONPATH
'''

from pysgpp import SerialCombiGrid,S_CT,DoubleVector

def f_3D(coords):
    return 1.0 + (0.25 * (coords[0] - 0.7) * (coords[0] - 0.7) + 2.0)\
            + (0.25 * (coords[1] - 0.7) * (coords[1] - 0.7) + 2.0)\
            + (0.25 * (coords[2] - 0.7) * (coords[2] - 0.7) + 2.0)



def main():
    dim=3
    level=3
    
    #setup combischeme
    scheme=S_CT(dim,level)
    
    #print subspaces
    levels=scheme.getLevels()
    coeffs=scheme.getCoef()
    for level,coef in zip(levels,coeffs):
        print level,coef
    
    #setup grid
    grid=SerialCombiGrid(scheme,True)
    
    #create full grids
    grid.createFullGrids()
    
    #fill data
    for i in range(grid.getNrFullGrid()):
        fgrid=grid.getFullGrid(i)
        data=fgrid.getElementVector()
        coord=DoubleVector(dim)
        for j in range(fgrid.getNrElements()):
            fgrid.getCoords(j,coord)
            data[j]=f_3D(coord)
    
    #evaluate the grid
    eval_coordinates=DoubleVector([0.1,0.1,0.1])
    print 'function value:\t',f_3D(eval_coordinates)
    print 'combined value:\t',grid.eval(eval_coordinates)

if __name__ == '__main__':
    main()

File: combigrid/examples/combi_classic.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#include <iostream>
#include "combigrid.hpp"

/*
 * The function to be interpolated
 */
double f_3D(std::vector<double> coords) {
  return 1.0 + (0.25 * (coords[0] - 0.7) * (coords[0] - 0.7) + 2.0)
         + (0.25 * (coords[1] - 0.7) * (coords[1] - 0.7) + 2.0)
         + (0.25 * (coords[2] - 0.7) * (coords[2] - 0.7) + 2.0);
}

using namespace combigrid;
int main() {

  int dim = 3;
  int level = 3;
  std::vector<double> coords(dim, 0.0);

  /*initialize combischeme (S_CT, TS_CT, CombiArbitraryScheme)*/
  CombiSchemeBasis* scheme = new S_CT(dim, level);

  /*get the combined subspaces and coefficients*/
  std::vector<std::vector<int> > levels = scheme->getLevels();
  std::vector<double> coeffs = scheme->getCoef();

  /*print the combined subspaces and coefficients*/
  std::cout << "combined subspaces: " << std::endl;

  for (size_t i = 0; i < coeffs.size(); i++) {
    for (size_t j = 0; j < levels[i].size(); j++) {
      std::cout << levels[i][j] << '\t';
    }

    std::cout << "|\t" << coeffs[i] << std::endl;
  }

  /*initialize combigrid (SerialCombiGrid, AdaptiveSerialCombiGrid)*/
  AbstractCombiGrid* grid = new SerialCombiGrid(scheme, true);

  /*create the respective full grids --> allocate the memory*/
  grid->createFullGrids();


  /*fill the array with the fullgrid data from the function*/
  for (int i = 0;
       i < grid->getNrFullGrid();
       i++
      ) {
    FullGridD* fgrid = grid->getFullGrid(i);
    std::vector<double> data = (fgrid->getElementVector());

    for (int j = 0; j < fgrid->getNrElements(); j++) {
      fgrid->getCoords(j, coords);

      fgrid->getElementVector()[j] = f_3D(coords);
    }
  }

  /*evaluate the interpolation by combination at a
   * certain coordinate.
   */
  std::cout.precision(8);
  std::vector<double> eval_coordinates(dim, 0.1);
  std::cout << "function value:\t" << f_3D(eval_coordinates)
            << std::endl;
  std::cout << "combined value:\t" << grid->eval(eval_coordinates)
            << std::endl;

}
